File: samengine-cli/src/linksaver.ts
#!/usr/bin/env node

// linksaver.ts
// TypeScript version of your Python Linksaver CLI

import { readFileSync, writeFileSync, existsSync } from "fs";
import { join } from "path";
import { exec } from "child_process";
import chalk from 'chalk';
import { mkdir } from "fs/promises";

const BASE_DIR = process.cwd();
const PATH = join(BASE_DIR, ".samengine/", "linksaver.json");

class ConfigError extends Error { }

// Types
export type Link = {
    name?: string;
    link: string;
    description: string;
    license?: string;
    author?: string;
    licenselink?: string;
};

// Config
export type AppConfig = {
    projectname: string;
    pretty: boolean;
    links: Link[];
};

// Function to generate a new Link
function newLink(link = "", description = "", license?: string, name?: string, author?: string, licenselink?: string): Link {
    const linkObj: Link = { link, description };

    if (license) {
        linkObj.license = license;
    }
    if (name) {
        linkObj.name = name;
    }
    if (author) {
        linkObj.author = author;
    }
    if (licenselink) {
        linkObj.licenselink = licenselink;
    }

    return linkObj;
}

function newAppConfig(projectname = ""): AppConfig {
    return {
        projectname,
        pretty: true,
        links: [],
    };
}

function requireString(data: any, key: string): string {
    const value = data[key];
    if (typeof value !== "string") {
        throw new Error(`${key} must be a string`);
    }
    return value;
}

function requireBool(data: any, key: string): boolean {
    const value = data[key];
    if (typeof value !== "boolean") {
        throw new Error(`${key} must be a boolean`);
    }
    return value;
}

function loadConfig(path: string): AppConfig {
    if (!existsSync(path)) {
        throw new ConfigError(`Config file not found: ${path}`);
    }

    let data: any;
    try {
        data = JSON.parse(readFileSync(path, "utf-8"));
    } catch (e) {
        throw new ConfigError(`Invalid JSON: ${e}`);
    }

    if (typeof data !== "object" || data === null) {
        throw new ConfigError("Config root must be an object");
    }

    try {
        if (!Array.isArray(data.links)) {
            throw new Error("links must be a list");
        }

        const links: Link[] = data.links.map((item: any, i: number) => {
            if (typeof item !== "object") {
                throw new Error(`links[${i}] must be an object`);
            }

            if (typeof item.link !== "string") {
                throw new Error(`links[${i}].link must be a string`);
            }

            if (typeof item.description !== "string") {
                throw new Error(`links[${i}].description must be a string`);
            }

            const linkObj: Link = {
                link: item.link,
                description: item.description,
            };

            if (typeof item.license === "string") {
                linkObj.license = item.license;
            }
            if (typeof item.name === "string") {
                linkObj.name = item.name;
            }
            if (typeof item.author === "string") {
                linkObj.author = item.author;
            }
            if (typeof item.licenselink === "string") {
                linkObj.licenselink = item.licenselink;
            }

            return linkObj;
        });

        return {
            projectname: requireString(data, "projectname"),
            pretty: requireBool(data, "pretty"),
            links,
        };
    } catch (e) {
        throw new ConfigError(`Invalid config structure: ${e}`);
    }
}

function saveConfig(config: AppConfig) {
    const indent = config.pretty ? 4 : undefined;

    writeFileSync(PATH, JSON.stringify(config, null, indent), "utf-8");
    console.log("Config saved.");
}

async function prompt(question: string): Promise<string> {
    const readline = await import("readline");

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    return new Promise((resolve) => {
        rl.question(question, (answer) => {
            rl.close();
            resolve(answer);
        });
    });
}

async function init() {
    console.log("Init Linksaver for Project");

    await mkdir(".webgameengine", { recursive: true });

    if (existsSync(PATH)) {
        console.log(`Config file already exists: ${PATH}`);
        return;
    }

    const projectname = await prompt("Projectname: ");
    const config = newAppConfig(projectname);

    saveConfig(config);
    console.log(`Created config file at ${PATH}`);
}

async function addLink(config: AppConfig) {
    const newName = await prompt("Name (optional): ");
    const newLinkVal = await prompt("New Link: ");
    const newDesc = await prompt("New Description: ");
    const newAuthor = await prompt("Author (optional): ");
    const newLic = await prompt("License (optional): ");
    const newLicenseLink = await prompt("License Link (optional): ");

    config.links.push(newLink(newLinkVal, newDesc, newLic, newName, newAuthor, newLicenseLink));
    console.log("Added new Link!");

    saveConfig(config);
}

function viewLinks(config: AppConfig): void {
    for (const link of config.links) {
        const parts = [
            `[${chalk.green(link.link)}]`,
            link.name ? `Name: ${link.name}` : null,
            `Desc: ${link.description}`,
            link.author ? `Author: ${link.author}` : null,
            link.license ? `License: ${chalk.red(link.license)}` : null,
            link.licenselink ? `License URL: ${link.licenselink}` : null,
        ].filter(Boolean);

        console.log(parts.join(" | "));
    }
}

function listLinks(config: AppConfig): void {
    for (const link of config.links) {
        let part = `"${link.name}" (${link.link}) by ${link.author} is licensed under ${link.license} (${link.licenselink})\n`;

        console.log(part);
    }
}

function openLink(link: string) {
    const platform = process.platform;
    let cmd = "";

    // Open Link on Windows
    if (platform === "win32") cmd = `start "" "${link}"`;
    // Open on Mac
    else if (platform === "darwin") cmd = `open "${link}"`;
    // Open on Linux and others
    else cmd = `xdg-open "${link}"`;

    exec(cmd, (err) => {
        if (err) console.error("Error opening link:", err);
    });
}

function openLinks(config: AppConfig) {
    console.log("Opening Links ...");
    for (const link of config.links) {
        openLink(link.link);
    }
}

function printhelp() {
    console.log(`
LINKSAVER CLI by webgameengine

Commands:
    help    show this message
    init    create a new Config
    add     add a new Link into the File
    view    view the links in the Terminal
    (none)  opening links in the browser
`);
}

// Main
// main.ts oder linksaver.ts
async function main() {
    // console.log("Linksaver TS");

    try {
        const args = process.argv.slice(2);

        if (args[0] === "-h" || args[0] === "--help" || args[0] === "help") {
            printhelp();
            return;
        }

        if (args[0] === "init") {
            await init();
            return;
        }

        let config: AppConfig;

        try {
            config = loadConfig(PATH);
        } catch (err) {
            if (err instanceof ConfigError) {
                console.error(chalk.red("⚠️  Config Error:", err.message));
                console.error("Please run 'init' to create a Config before.");
                process.exit(1);
            } else {
                throw err; // alles andere weiter werfen
            }
        }

        if (args[0] === "add") {
            await addLink(config);
            return;
        }

        if (args[0] === "view") {
            viewLinks(config);
            return;
        }

        if (args[0] === "list") {
            listLinks(config);
            return;
        }

        openLinks(config);

    } catch (err) {
        console.error("Unerwarteter Fehler:", err);
        process.exit(1);
    }
}

await main();

File: tool/src/linksaver.rs
use serde::{Deserialize, Serialize};
use std::{
    env, fs,
    io::{self, Write},
    path::PathBuf,
    process::Command,
};

#[derive(Serialize, Deserialize, Debug, Clone)]
struct Link {
    name: Option<String>,
    link: String,
    description: String,
    license: Option<String>,
    author: Option<String>,
    licenselink: Option<String>,
    showinlist: bool,
    changenotice: bool,
}

#[derive(Serialize, Deserialize, Debug)]
struct AppConfig {
    projectname: String,
    pretty: bool,
    links: Vec<Link>,
    links2: Vec<String>,
}

// ---------- PATH ----------

fn config_path() -> PathBuf {
    let mut path = env::current_dir().unwrap();
    path.push(".samengine/linksaver.json");
    path
}

// ---------- IO ----------

fn prompt(msg: &str) -> String {
    print!("{}", msg);
    io::stdout().flush().unwrap();

    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}

// ---------- CONFIG ----------

fn new_config(name: String) -> AppConfig {
    AppConfig {
        projectname: name,
        pretty: true,
        links: vec![],
        links2: vec![],
    }
}

fn save(config: &AppConfig) -> Result<(), std::io::Error> {
    let path = config_path();

    let json = if config.pretty {
        serde_json::to_string_pretty(config).unwrap()
    } else {
        serde_json::to_string(config).unwrap()
    };

    fs::write(path, json)?;
    Ok(())
}

fn load() -> Result<AppConfig, String> {
    let path = config_path();

    if !path.exists() {
        return Err("config not found".into());
    }

    let data = fs::read_to_string(path).map_err(|e| e.to_string())?;
    let config: AppConfig =
        serde_json::from_str(&data).map_err(|e| format!("invalid json: {e}"))?;

    if config.projectname.is_empty() {
        return Err("projectname must be set".into());
    }

    Ok(config)
}

// ---------- COMMANDS ----------

fn init() {
    println!("Init Linksaver");

    let dir = env::current_dir().unwrap().join(".samengine");
    fs::create_dir_all(&dir).unwrap();

    let path = config_path();
    if path.exists() {
        println!("Config already exists: {:?}", path);
        return;
    }

    let name = prompt("Projectname: ");
    let config = new_config(name);

    save(&config).unwrap();
    println!("Created config at {:?}", path);
}

fn add(config: &mut AppConfig) {
    let link = Link {
        name: {
            let v = prompt("Name (optional): ");
            if v.is_empty() { None } else { Some(v) }
        },
        link: prompt("New Link: "),
        description: prompt("New Description: "),
        author: {
            let v = prompt("Author (optional): ");
            if v.is_empty() { None } else { Some(v) }
        },
        license: {
            let v = prompt("License (optional): ");
            if v.is_empty() { None } else { Some(v) }
        },
        licenselink: {
            let v = prompt("License Link (optional): ");
            if v.is_empty() { None } else { Some(v) }
        },
        showinlist: prompt("Show in list? (y/n, default y): ") != "n",
        changenotice: prompt("Mark as changed? (y/n, default n): ") == "y",
    };

    config.links.push(link);
    save(config).unwrap();
    println!("Added new link!");
}

fn add2(config: &mut AppConfig) {
    let entry = prompt("Entry text: ");
    config.links2.push(entry);
    save(config).unwrap();
    println!("Added new entry!");
}

fn view(config: &AppConfig) {
    for l in &config.links {
        print!("[{}] ", l.link);

        if let Some(name) = &l.name {
            print!("Name: {} | ", name);
        }

        print!("Desc: {} | ", l.description);

        if let Some(author) = &l.author {
            print!("Author: {} | ", author);
        }

        if let Some(lic) = &l.license {
            print!("License: {} | ", lic);
        }

        if let Some(ll) = &l.licenselink {
            print!("License URL: {}", ll);
        }

        if l.changenotice {
            print!(" (changes were made)");
        }

        println!();
    }
}

fn list(config: &AppConfig) {
    println!("\nCredits:\n");

    for l in &config.links {
        if !l.showinlist {
            continue;
        }

        println!(
            "\"{}\" ({}) by {} is licensed under {} ({}){}",
            l.name.clone().unwrap_or_default(),
            l.link,
            l.author.clone().unwrap_or_default(),
            l.license.clone().unwrap_or_default(),
            l.licenselink.clone().unwrap_or_default(),
            if l.changenotice {
                " (changes were made)"
            } else {
                ""
            }
        );
    }

    for e in &config.links2 {
        println!("{}", e);
    }
}

fn open_link(url: &str) {
    let cmd = if cfg!(windows) {
        Command::new("cmd").args(["/C", "start", url]).spawn()
    } else if cfg!(target_os = "macos") {
        Command::new("open").arg(url).spawn()
    } else {
        Command::new("xdg-open").arg(url).spawn()
    };

    if let Err(e) = cmd {
        eprintln!("Error opening link: {}", e);
    }
}

fn open_all(config: &AppConfig) {
    println!("Opening links...");
    for l in &config.links {
        open_link(&l.link);
    }
}

// ---------- HELP ----------

fn help() {
    println!(
        r#"
LINKSAVER CLI

Commands:
    help    show this message
    init    create config
    add     add link
    add2    add entry (text only)
    view    view links
    list    list links
    (none)  open all links
"#
    );
}

// ---------- MAIN ----------

pub fn execute(arg: &str) {
    match arg {
        "help" | "-h" | "--help" => {
            help();
            return;
        }
        "init" => {
            init();
            return;
        }
        _ => {}
    }

    let mut config = match load() {
        Ok(c) => c,
        Err(e) => {
            eprintln!("Linksaver: Config Error: {}", e);
            eprintln!("Run 'init' first.");
            std::process::exit(1);
        }
    };

    match arg {
        "add" => add(&mut config),
        "add2" => add2(&mut config),
        "view" => view(&config),
        "list" => list(&config),
        _ => open_all(&config),
    }
}
